pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +106,243 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToConvertMachineSet", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to convert CAPI MachineSet to MAPI MachineSet: %w", err)
+	}
+
+	for _, warning := range warns {
+		logger.Info("Warning during conversion", "warning", warning)
+	}
+
+	// Use a copy of the MAPI MachineSet for any changes as the helper functions
+	// update in place
+	updatedMapiMachineSet := mapiMachineSet.DeepCopy()
+
+	// Update the MAPI MachineSet spec
+	updateMAPIMachineSetSpecFromCAPI(ctx, updatedMapiMachineSet, capiMachineSet,
+		convertedMachineSet.Spec.Template.Spec.ProviderSpec)
+
+	// Check if there are any changes after updating the spec
+	if !reflect.DeepEqual(updatedMapiMachineSet.Spec, mapiMachineSet.Spec) {
+		logger.Info("Updating MAPI MachineSet spec")
+
+		if err := r.Update(ctx, updatedMapiMachineSet); err != nil {
+			logger.Error(err, "Failed to update MAPI MachineSet")
+
+			r.updateSynchronisedConditionWithPatch(
+				ctx, mapiMachineSet, corev1.ConditionFalse,
+				"FailedToUpdateMAPI", err.Error(),
+			)
+
+			return ctrl.Result{}, fmt.Errorf("failed to update MAPI MachineSet: %w", err)
+		}
+
+		logger.Info("Successfully updated MAPI MachineSet spec")
+	} else {
+		logger.Info("No changes detected in MAPI MachineSet spec")
+	}
+
+	// Update the MAPI MachineSet status
+	updateMAPIMachineSetStatusFromCAPI(ctx, updatedMapiMachineSet, capiMachineSet)
+
+	// Check if there are any changes after updating the status
+	if !reflect.DeepEqual(updatedMapiMachineSet.Status, mapiMachineSet.Status) {
+		logger.Info("Updating MAPI MachineSet status")
+
+		if err := r.Status().Update(ctx, updatedMapiMachineSet); err != nil {
+			logger.Error(err, "Failed to update MAPI MachineSet status")
+
+			r.updateSynchronisedConditionWithPatch(
+				ctx, mapiMachineSet, corev1.ConditionFalse,
+				"FailedToUpdateMAPI", err.Error(),
+			)
+
+			return ctrl.Result{}, fmt.Errorf("failed to update MAPI MachineSet status: %w", err)
+		}
+
+		logger.Info("Successfully updated MAPI MachineSet status")
+	} else {
+		logger.Info("No changes detected in MAPI MachineSet status")
+	}
+
+	r.updateSynchronisedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionTrue,
+		consts.ReasonResourceSynchronised, "Synchronised CAPI to MAPI")
+
 	return ctrl.Result{}, nil
 }
 
+func (r *MachineSetSyncReconciler) updateSynchronisedConditionWithPatch(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, status corev1.ConditionStatus, reason, message string) {
+	logger := log.FromContext(ctx).WithValues("function", "updateSynchronisedConditionWithPatch")
+
+	oldMachineSet := mapiMachineSet.DeepCopy()
+
+	var newCondition *machinev1beta1.Condition
+	if status == corev1.ConditionTrue {
+		newCondition = conditions.TrueConditionWithReason(
+			consts.SynchronisedCondition, reason, message, //nolint:govet
+		)
+
+		// Update the synchronizedGeneration
+		mapiMachineSet.Status.SynchronizedGeneration++
```

Comment:
The generation should come from the generation of the source machineset, it's not an internal counter